### PR TITLE
Version 1 with backend and testing

### PR DESCRIPTION
**Pull request by Rupamdas832**
-Login/signup
-Result and review page
-Leaderboard for all users

### CODE REVIEW

File: src/Tests/quizReducer.test.ts [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                isCorrect: false,
+              },
+            ],
+          },
+        ],
+        highestScore: 0,
+      },
+    };
+
+    let state = QuizReducer(initialState, <Action>action);
```
Comment (tanaypratap) [inline src/Tests/quizReducer.test.ts]: const instead...

File: src/Tests/quizReducer.test.ts [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+      totalQuestions: 6,
+      correctAnswers: 0,
+      accuracy: 0,
+      highestScore: 0,
+      questionsAttempted: [],
+    };
+    const finalState = {
+      _id: "",
+      score: 0,
+      status: "starting",
```
Comment (tanaypratap) [inline src/Tests/quizReducer.test.ts]: hey Rupam, looks like you haven't updated the status.

File: src/App.tsx [hunk 3]
```diff
@@ -1,13 +1,46 @@
...
+  const { storeState, storeDispatch } = useStore();
+  const { loadingMessage } = storeState;
+
+  const fetchQuizzes = async () => {
+    storeDispatch({ type: "IS_LOADING", payload: "loading" });
+    try {
+      const response = await axios(`${URL}/quiz`);
+      const quizzes = response.data.quizzes;
+      if ((response.status = 200)) {
+        storeDispatch({ type: "LOAD_QUIZZES", payload: quizzes });
```
Comment (DanishOnGit) [inline src/App.tsx]: we can destructure the response  `const {data:{quizzes},status} = await axios.get...`

File: src/Components/PrivateRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import { Route, Navigate } from "react-router-dom";
+
+export const PrivateRoute = ({ ...props }) => {
+  const storage = localStorage?.getItem("QuizLoginUser");
+
+  if (storage !== null) {
+    const loginStatus = JSON.parse(storage);
+    if (loginStatus.isUserLogin === true) {
```
Comment (DanishOnGit) [inline src/Components/PrivateRoute.tsx]: I think `if (loginStatus.isUserLogin){...}` will work the same.

File: src/Pages/Login.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  const { userDispatch } = useUser();
+
+  const { storeState, storeDispatch } = useStore();
+  const { loadingMessage } = storeState;
+
+  const loginWithCredentials = async () => {
+    storeDispatch({ type: "IS_LOADING", payload: "loggingIn" });
+    try {
+      const response = await axios.post(`${URL}/login`, {
```
Comment (DanishOnGit) [inline src/Pages/Login.tsx]: here too we can destructure in same manner as before

File: src/Components/HeroSection.tsx [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import { QuestionCard } from "../Components";
+import { PrivateRoute } from "./PrivateRoute";
+
+export const HeroSection = () => {
+  return (
+    <div>
+      <Routes>
+        <Route path="/" element={<Home />} />
+        <Route path="/categories" element={<Categories />} />
+        <PrivateRoute path="/quiz" element={<QuestionCard />} />
```
Comment (poojasabnani) [inline src/Components/HeroSection.tsx:21]: Private routes can be added together

File: src/Components/QuestionCard.tsx [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+import { Link, useNavigate } from "react-router-dom";
+import { Option, Question } from "../Data/quizData.types";
+import { useQuiz, useStore, useUser } from "../Store";
+import URL from "./ServerURL";
+import { Toast } from "./Toast";
+
+export const QuestionCard = () => {
+  const [isAnswered, setIsAnswered] = useState(false);
+  const [timeCounter, setTimeCounter] = useState(20);
+  const [isSkipped, setIsSkipped] = useState(false);
```
Comment (poojasabnani) [inline src/Components/QuestionCard.tsx]: possible to use useReducer here as 3 states being used.

File: src/Pages/Signup.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import { useStore, useUser } from "../Store";
+import axios from "axios";
+import URL from "../Components/ServerURL";
+import { Toast } from "../Components";
+
+export const Signup = () => {
+  const [name, setName] = useState("");
+  const [email, setEmail] = useState("");
+  const [password, setPassword] = useState("");
+  const [error, setError] = useState("");
```
Comment (poojasabnani) [inline src/Pages/Signup.tsx:15]: here also a useReducer can be used

File: src/Pages/Signup.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+  const { storeState, storeDispatch } = useStore();
+  const { loadingMessage } = storeState;
+
+  const { userDispatch } = useUser();
+
+  const signUpUser = async () => {
+    storeDispatch({ type: "IS_LOADING", payload: "signup" });
+    try {
+      const response = await axios.post(`${URL}/signup`, {
```
Comment (poojasabnani) [inline src/Pages/Signup.tsx]: response can be destructured

**Review by poojasabnani — COMMENTED** [review]
Hey Rupam, very well structured and nicely detailed written code. I personally learnt a lot reviewing your code.
I've suggested some minor changes. Do have a look.

File: src/Components/HeroSection.tsx (hunk 6 above)
Comment (Rupamdas832) [reply src/Components/HeroSection.tsx:21]: Sure I will do.

File: src/Components/QuestionCard.tsx (hunk 7 above)
Comment (Rupamdas832) [reply src/Components/QuestionCard.tsx]: Good point. Thanks, I will refactored it to a reducer.

**Comment by Rupamdas832** [comment]
> Hey Rupam, very well structured and nicely detailed written code. I personally learnt a lot reviewing your code.
> I've suggested some minor changes. Do have a look.

Glad to hear that you learned something from the code.

File: src/App.tsx (hunk 3 above)
Comment (Rupamdas832) [reply src/App.tsx]: Done.

File: src/Components/PrivateRoute.tsx (hunk 4 above)
Comment (Rupamdas832) [reply src/Components/PrivateRoute.tsx]: Done

File: src/Pages/Login.tsx (hunk 5 above)
Comment (Rupamdas832) [reply src/Pages/Login.tsx]: Done

File: src/Pages/Signup.tsx (hunk 9 above)
Comment (Rupamdas832) [reply src/Pages/Signup.tsx]: Done

File: src/Tests/quizReducer.test.ts (hunk 1 above)
Comment (Rupamdas832) [reply src/Tests/quizReducer.test.ts]: Done

File: src/Tests/quizReducer.test.ts (hunk 2 above)
Comment (Rupamdas832) [reply src/Tests/quizReducer.test.ts]: Removed unused state

**Pull request merged**